File: src/main/python/recur_number_of_cases/퇴사14501.py
N = int(input())

input_list = [list(map(int, input().split())) for _ in range(N)]
answer = 0
def recur(t,p):
  global answer
  if t > N-1:
    if t > N:return
    answer = max(answer,p)
    return

  recur(t+input_list[t][0],p+input_list[t][1])
  recur(t+1,p)

recur(0,0)
print(answer)

File: src/main/java/org/snust/recur_number_of_cases/퇴사14501.java
package org.snust.recur_number_of_cases;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class 퇴사14501 {
  private static int answer = Integer.MIN_VALUE;
  private static int N;
  private static int[][]  inputList;
  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    N = Integer.parseInt(br.readLine());
    StringTokenizer st;
    inputList = new int[N][2];
    for (int i = 0; i < N; i++) {
      st = new StringTokenizer(br.readLine(), " ");
      inputList[i][0] = Integer.parseInt(st.nextToken());
      inputList[i][1] = Integer.parseInt(st.nextToken());
    }
    recur(0, 0);
    System.out.println(answer);
  }

  private static void recur(int t, int p) {
    if (t > N-1) {
      if (t > N) {
        return;
      }
      answer = Math.max(answer, p);
      return;
    }
    recur(t + inputList[t][0], p+inputList[t][1]);
    recur(t + 1, p);
  }
}
